### ISSUE TITLE
Invoke a route outside a React component

### ISSUE THREAD

**Issue by raypatterson** [STRML/react-router-component#105]
Hi, 

Great work on this library!

I'm wondering if there is a way to invoke a route without the `Router.NavigatableMixin` in a React component or the `Link` component. 

I am using a Flux lib, passing a value to a store, doing an async request and based on the response I'd like to update the React views via a URL path change.

Is this a possibility?

Thanks!

**Comment by jurassix** [comment]
+1

**Comment by jurassix** [comment]
@RayPatterson I've skirted this issue with the following pattern. 

Overall, I think it's the correct way to handle this within this library, but it is still not fully ideal. I'd like to be able to simply require in the react-router-component and dispatch a new route statically from my ActionCreators. Until a static navigate option is available from this library, the bellow pattern is working for me:

First I have a LocationStore that allows Flux Actions to dispatch new routes via Stores. Simple handle specific dispatched events and update the LocationStore's path with the desired new route and emit a change:

```
var assign = Object.assign || require('object.assign');
var EventEmitter = require('events').EventEmitter;
var AppDispatcher = require('../dispatcher/AppDispatcher');
var ActionTypes = require('../constants/ActionTypes');
var BaseConstants = require('../constants/BaseConstants');
var SessionStore = require('../../user/stores/SessionStore');

var LocationStore = assign({
  path: '/'
}, EventEmitter.prototype, {
  emitChange: function() {
    return this.emit(BaseConstants.CHANGE_EVENT);
  },
  addChangeListener: function(callback) {
    return this.on(BaseConstants.CHANGE_EVENT, callback);
  },
  removeChangeListener: function(callback) {
    return this.removeListener(BaseConstants.CHANGE_EVENT, callback);
  },
  setPath: function(path) {
    this.path = path;
  },
  getPath: function() {
    return this.path || '/';
  }
});

LocationStore.dispatchToken = AppDispatcher.register(function(payload) {
  switch (payload.type) {
    case ActionTypes.SIGNIN_SUCCESSFUL:
    case ActionTypes.SIGNUP_SUCCESSFUL:
      AppDispatcher.waitFor([SessionStore.dispatchToken]);
      LocationStore.setPath('/lists');
      break;
    default:
      return;
  }
  LocationStore.emitChange();
});

module.exports = LocationStore;
```

Next, I simply add the following JSX component to at the same level as my main router. This Component simply listens to LocationStore events and navigates, via react-router-component, to the dispatched path:

```
var React = require('react');
var NavigatableMixin = require('react-router-component').NavigatableMixin
var LocationStore = require('../stores/LocationStore');

var FluxRoutes = React.createClass({
  displayName: 'FluxRoutes',
  mixins: [NavigatableMixin],
  componentWillMount: function() {
    LocationStore.addChangeListener(this.onLocation);
  },
  componentWillUnmount: function() {
    LocationStore.removeChangeListener(this.onLocation);
  },
  render: function() {
    return null;
  },
  onLocation: function() {
    this.navigate(LocationStore.getPath());
  }
});

module.exports = FluxRoutes;
```

Finally, if I need to restrict certain routes based on authentication, I use render a different set of routes for users not logged in verse authenticated users. 

Hope this helps. 

Let me know how you have handled this issue if you have a better pattern.

**Comment by RayPatterson** [comment]
@jurassix I ended up trying https://github.com/larrymyers/react-mini-router as it has a `navigate` method. Worked well enough for my needs.

**Comment by RayPatterson** [comment]
@jurassix I ended up switching back as my route pattens became a little more complex than what splats could support.

My hack solution was to mount this component at the top of my app, expose the `navigate` method and require it in my stores.

```
var React = require('react');
var Link = require('react-router-component').Link;

var _navigate;

module.exports = {

  Component: React.createClass({

    componentDidMount: function() {

      _navigate = this.refs.el.navigate;
    },

    render: function() {
      return (
        <Link ref="el" href="" />
      );
    }
  }),

  navigate: _navigate
};
```
[I _think_ I'm following the direction given in the docs.](http://strml.viewdocs.io/react-router-component)

> Alternatively, if you have a reference to a router component instance, you can call its .navigate(href) method to do a transition to a different location. You can acquire a reference to a router by using React's Refs mechanism.

**Comment by STRML** [comment]
Yep, that works. You can navigate via any <Link> or <Locations>/<Pages>
instance.

**Comment by RayPatterson** [comment]
@STRML Thanks!

**Comment by jurassix** [comment]
I have reservations about referencing a React component from outside of React. I will stick to dispatching events and having a React component register to those events. 

Keeping a reference to a React component outside of React will potentially lead to invariant issues.

**Comment by RayPatterson** [comment]
@jurassix My solution is definitely a "path of least resistance" workaround but since I'm working within a personal project I have the luxury of being lazy :)

I think I understand what you mean about the dangers of the external reference but if you could elaborate that would be helpful. I assume you mean that the reference would be lost if the parent component was unmounted?

Or maybe there are deeper concerns?

**Comment by jurassix** [comment]
@RayPatterson Mainly my concern comes from advocating a pattern that would rely on using a React component outside of React. Unless you are referencing a React components `static` block, you have no guarantees about the React component reference your passing around. Your app is OK because you have mounted this component at the Root. However, other developers may see this as a pattern and not implicitly know that this component has this dependency and may try to blindly copy it. 

When I first started with FLUX I naively tried attaching callbacks to actions which caused invariant issues if I tried to update state etc. I then added stores and listeners and the need for callbacks was gone. This pretty much burnt me on referencing React components outside of React. 

So my deeper concerns are more for the next guy to work with your code than with anything else.
